### PR TITLE
type_question_answer

### PR DESCRIPTION
**Pull request by shayb**
support new type of questionnaire -  quiz

### CODE REVIEW

File: app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt [hunk 1]
```diff
@@ -64,6 +63,309 @@ class TaskService(context: Context, api: TasksApi,
...
+  },
+  "min_to_complete": 2,
+  "memo": "1-kit-sb0ea956eabef4705bd26b",
+  "tags": [
+    "Finance"
+  ],
+  "type": "quiz",
+  "title": "Kin Quiz",
+  "price": 50,
+  "desc": "Let's test your Kinowledge!",
```
Comment (hello8721) [inline app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: spelling

File: app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt [hunk 2]
```diff
@@ -64,6 +63,309 @@ class TaskService(context: Context, api: TasksApi,
...
+    {
+      "id": "844",
+      "results": [
+        {
+          "id": "46584",
+          "text": "I know that already (Correct)"
+        },
+{
+          "id": "46585",
+          "text": "wrong I know that already"
```
Comment (hello8721) [inline app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: capitalize

File: app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt [hunk 3]
```diff
@@ -64,6 +63,309 @@ class TaskService(context: Context, api: TasksApi,
...
+        {
+          "id": "4589",
+          "text": "I'm still a bit confused about that"
+        }
+      ],
+      "type": "text",
+      "text": "Kin is a cryptocurrency that can be earned or spent on things you like.",
+      "quiz_data": {
+        "answer_id": "46584",
+        "explanation": "Kin is a cryptocurrency that can be earned or spent on things you like ecosystem of digital services.",
```
Comment (hello8721) [inline app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: explanation makes no sense

File: app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt [hunk 4]
```diff
@@ -64,6 +63,309 @@ class TaskService(context: Context, api: TasksApi,
...
+        "explanation": "you getting Kin from God",
+        "reward": 5
+      }
+    },
+    {
+      "id": "372",
+      "results": [
+{
+          "id": "342756",
+          "text": "wrobnf Got it!"
```
Comment (hello8721) [inline app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: spelling

Comment (fuzzyami) [reply app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: @hello8721 thanks for your review! However, I believe what you're seeing is just temporary, throw-away code. this content will eventually come from the server and will not be hard coded in the client. its just until the feature is supported in the server (read: until I do my work :] )

File: app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt (hunk 1 above)
Comment (fuzzyami) [reply app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: probably intentional. kin-knowledge. get it?

File: app/src/main/java/org/kinecosystem/kinit/model/earn/Question.kt [hunk 5]
```diff
@@ -21,6 +23,13 @@ data class Question(val id: String? = null,
...
 
+data class QuizData(@SerializedName("answer_id")
+                    val answer_id: String,
+                    @SerializedName("explanation")
+                    val explanation: String,
+                    @SerializedName("reward")
+                    val reward: Int)
+
 fun Question.isValid(): Boolean {
     if (id.isNullOrBlank() || text.isNullOrBlank() || type.isNullOrBlank() || answers.orEmpty().isEmpty()) {
```
Comment (katzio) [inline app/src/main/java/org/kinecosystem/kinit/model/earn/Question.kt]: No validation for quiz_data (if not `null`)?

File: app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt (hunk 4 above)
Comment (berryve) [reply app/src/main/java/org/kinecosystem/kinit/network/TaskService.kt]: @shayb this shouldn't be merged to master

File: app/src/main/java/org/kinecosystem/kinit/navigation/Navigator.kt [hunk 6]
```diff
@@ -58,10 +57,10 @@ class Navigator(private val context: Context) {
 
     private fun navigateToTask() {
         if (tasksRepository.task != null) {
```
Comment (mrsegev) [inline app/src/main/java/org/kinecosystem/kinit/navigation/Navigator.kt]: Could be converted to `let` statement.

File: app/src/main/java/org/kinecosystem/kinit/view/customView/QuizAnswerView.kt [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+                animateBackground(true)
+            }
+        }
+    }
+
+    private fun animateReward() {
+        reward_text.y = reward_text.height.toFloat()
+        reward_text.alpha = 1f
+        reward_text.animate().setStartDelay(1100).translationY(0f).setInterpolator(OvershootInterpolator(3f)).setDuration(600).withEndAction {
+            answer_text.postDelayed({ listener.onAnimComplete() }, 2000)
```
Comment (mrsegev) [inline app/src/main/java/org/kinecosystem/kinit/view/customView/QuizAnswerView.kt]: Might cause problems if the user navigates away before the time is up (2 seconds).
I would check if the listener exists before invoking `onAnumComplete()` and add option to remove the listener according to the appropriate lifecycle.

File: app/src/main/java/org/kinecosystem/kinit/view/earn/AnswerExplainedFragment.kt [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+
+        val binding = DataBindingUtil.inflate<AnswerExplainedFragmentBinding>(inflater, R.layout.answer_explained_fragment, container, false)
+        if (activity !is QuestionnaireActions) {
+            onInvalidData()
+        }
+        if (arguments == null || !arguments!!.containsKey(ARG_QUIZ_INDEX)) {
+            onInvalidData()
+        }
+        val questionIndex = arguments!!.getInt(ARG_QUIZ_INDEX, INVALID_QUIZ_INDEX)
```
Comment (mrsegev) [inline app/src/main/java/org/kinecosystem/kinit/view/earn/AnswerExplainedFragment.kt]: arguments might be null

File: app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+    var showCorrect: ObservableBoolean = ObservableBoolean(false)
+
+    var selectionListener: QuizAnswerView.OnSelectionListener = object : QuizAnswerView.OnSelectionListener {
+        override fun onAnimComplete() {
+            questionnaireActions.next()
+        }
+
+        override fun onAnswered(answer: Answer?) {
+            clickable.set(false)
+            answer?.let {
```
Comment (mrsegev) [inline app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt]: Why Answer's id is an optional field?

File: app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+    fun isCorrect(answer: Answer?): Boolean {
+        answer?.let {
+            return questionObj?.quiz_data?.answer_id == it.id
+        }
+        return false
+    }
+
+    fun onNext() {
+        val answerId = questionObj?.id ?: ""
```
Comment (mrsegev) [inline app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt]: Does the code know how to deal with an empty string as id? Why not use nulls?

File: app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+            val task = taskRepository.task
+            taskRepository.setChosenAnswers(answerId, chosenAnswers)
+            questionnaireActions.next()
+            analytics.logEvent(answerEvent(task))
+        }
+    }
+
+    val onAnswer = object : AnswersListAdapter.OnAnswerSelectedListener {
+        override fun answerSelected(answer: Answer?): Boolean {
+            val answerId = answer?.id ?: ""
```
Comment (mrsegev) [inline app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt]: Same as above

File: app/src/main/java/org/kinecosystem/kinit/model/earn/Question.kt (hunk 5 above)
Comment (shayb) [reply app/src/main/java/org/kinecosystem/kinit/model/earn/Question.kt]: @berryve done

File: app/src/main/java/org/kinecosystem/kinit/view/earn/AnswerExplainedFragment.kt (hunk 8 above)
Comment (shayb) [reply app/src/main/java/org/kinecosystem/kinit/view/earn/AnswerExplainedFragment.kt]: there is check in line 26

File: app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt (hunk 9 above)
Comment (shayb) [reply app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt]: answer has a custom set hence it cant be lateinit

Comment (shayb) [reply app/src/main/java/org/kinecosystem/kinit/viewmodel/earn/QuizQuestionViewModel.kt]: if theres error in the data taken from server we dont want to crash in the retrofit parsing
so data class from server might have null and we check if its valid

File: app/src/main/java/org/kinecosystem/kinit/view/customView/QuizAnswerView.kt (hunk 7 above)
Comment (shayb) [reply app/src/main/java/org/kinecosystem/kinit/view/customView/QuizAnswerView.kt]: i tried to test it make the animation long and navigate backed or keep in background it worked as expected

**Pull request merged**